### PR TITLE
Tong Guan

### PR DESCRIPTION
**Pull request by YuanYuanBingZi**

### CODE REVIEW

File: brevets/acp_times.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+               minutes += minute_time_reference[min_km]
+               if minutes > 60:
+                   hours += minutes // 60
+                   minutes = minutes % 60
+
+    
+    return brevet_start_time.shift(hours=hours, minutes=minutes)
+
+
+def close_time(control_dist_km, brevet_dist_km, brevet_start_time):
```
Comment (ZackJ94) [inline brevets/acp_times.py:57]: Just like in open_time(), I think your solution here is very clever, and would be more easily adapted to an expanded calculator than it would have been if you just did hard checks for each distance "bracket."

File: brevets/flask_brevets.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#
+###############
+@app.route("/_calc_times")
+def _calc_times():
+    """
+    Calculates open/close times from miles, using rules
+    described at https://rusa.org/octime_alg.html.
+    Expects one URL-encoded argument, the number of miles.
+    """
+    app.logger.debug("Got a JSON request")
```
Comment (ZackJ94) [inline brevets/flask_brevets.py:52]: This looks good. Very similar to how I got the info in my project. 
I would suggest that, once you've completed the task, you delete the FIXME comments so its clear that the logic has been implemented.

File: brevets/templates/calc.html [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                     var controls = []
+
+                     $('.control').each(function(){
+                        var checkpoint = $(this);
+
+                        var km = checkpoint.find("input[name='km']").val();
+                        var open_time_field = checkpoint.find("input[name='open']");
+                        var close_time_field = checkpoint.find("input[name='close']");
+
+                        if(km){
```
Comment (ZackJ94) [inline brevets/templates/calc.html:166]: This is a good solution, but it doesn't submit the miles or optional location info for each control.

File: brevets/tests/test_acp_times.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        300:(start_time.shift(hours = 9), start_time.shift(hours = 20))
+    }
+
+    for km, time_tuple in checkpoints.items():
+        check_open, check_close = time_tuple
+        assert open_time(km, dist, start_time) == check_open
+        assert close_time(km, dist, start_time) == check_close
+
+
+def test_brevet3():
```
Comment (ZackJ94) [inline brevets/tests/test_acp_times.py:48]: Not really an issue, but it looks like test_brevet3() made it into the test suite twice.

File: brevets/tests/test_db.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        200:(start_time.shift(hours = 5, minutes = 53), start_time.shift(hours = 13, minutes = 30))
+    }
+
+    id = insert_brevet(dist, start_time, checkpoints)
+    new_start_time, new_dist, new_checkpoints = get_brevet()
+    assert new_start_time == start_time
+    assert new_dist == dist
+    assert new_checkpoints == checkpoints
+
+
```
Comment (ZackJ94) [inline brevets/tests/test_db.py]: The test for retrieving a brevet seems to be missing.

File: README.md [hunk 6]
```diff
@@ -10,20 +10,31 @@ Presently, there's only a placeholder directory for your Flask app, and a `docke
...
+	* Web app instruction
+	  * First, you need to choose the end km or the length of the race based on the options
+	  * Second, you need to choose the date you want to query
+	  * Third, you can type the checkpoint km one by one and press return, you will get the exact time for the open time and close time.
+	  * Fourth, you can press the submit button to submit all the data to the database. After submission, when you press display buttion, you will get all the data back.
 
-3. Update README.md with brevet control time calculation rules (you were supposed to do this for Project 4), and additional information regarding this project.
-	- This project will be peer-reviewed, so be thorough.
 
 ## Grading Rubric
```
Comment (ZackJ94) [inline README.md:39]: the README is well done. I think you went into a good amount of detail and I was able to get the app running with no issues.

**Review by ZackJ94 — COMMENTED** [review]
Your project is very well done!

I only found a couple small issues during testing:
1. Miles and location are not submitted to / fetched from the database.

2. The values in the webpage fields are not cleared once the data is submitted to the database. This could be solved by setting their values to an empty string (i.e., field.val("")).

3. Your tests for acp_times could be a bit more robust to check different edge cases, but I didn't find any bugs during testing so I believe the tests you have currently are adequate.

The rest of the files all look good, and you followed the instruction and skelton code well. Great job!
